### PR TITLE
Concurrent tasks

### PR DESCRIPTION
**Pull request by andrewMacmurray**
This adds concurrent tasks to gren. The existing `Task` api has been modified to:

- Add `Task.concurrent`, which runs an array of tasks concurrently.
- Make `map2` run concurrently.
- Remove `map4`, `map5` in favour of `andMap`.

The existing `Scheduler` kernel implementation is largely unchanged but with the addition of a `_Scheduler_concurrent` helper that spawns multiple processes, handles collecting results and signalling errors.

### CODE REVIEW

File: src/Gren/Kernel/Scheduler.js [hunk 1]
```diff
@@ -49,24 +49,77 @@ function _Scheduler_receive(callback) {
   };
 }
 
+function _Scheduler_concurrent(tasks) {
+  return _Scheduler_binding(function (callback) {
+    if (tasks.length === 0) return callback(_Scheduler_succeed([]));
```
Comment (robinheghan) [inline src/Gren/Kernel/Scheduler.js:54]: Don't think you have to have this inside the binding. Simply return success if `tasks.length === 0`.

File: src/Gren/Kernel/Scheduler.js [hunk 2]
```diff
@@ -113,6 +170,18 @@ type alias Process =
 var _Scheduler_working = false;
 var _Scheduler_queue = [];
 
+function _Scheduler_enqueueBatch(procs) {
```
Comment (robinheghan) [inline src/Gren/Kernel/Scheduler.js:173]: I don't think the implementation here is entirely correct, as you're not checking if `_Scheduler_working` is already true here.

That said, I believe a better approach is to move the "execute queue if not already executing" code to a new function, and then decide how to register tasks (single or batch) before calling that function.

Something like:

```js
function _Scheduler_enqueue(proc) {
  _Scheduler_queue.push(proc);
  _Scheduler_execute();
}

function _Scheduler_enqueueBatch(procs) {
  for (const proc of procs) {
    _Scheduler_queue.push(proc);
  }
  _Scheduler_execute();
}

function _Scheduler_execute() {
  if (_Scheduler_working) {
    return;
  }
  _Scheduler_working = true;
  while ((proc = _Scheduler_queue.shift())) {
    _Scheduler_step(proc);
  }
  _Scheduler_working = false;
}
```

File: src/Task.gren [hunk 3]
```diff
@@ -145,70 +150,38 @@ If it fails, the whole thing fails!
...
+    succeed func
+        |> andMap taskA
+        |> andMap taskB
+        |> andMap taskC
 
 
 {-| -}
 map4 : (a -> b -> c -> d -> result) -> Task x a -> Task x b -> Task x c -> Task x d -> Task x result
 map4 func taskA taskB taskC taskD =
-    taskA
```
Comment (robinheghan) [inline src/Task.gren:175]: Since we're introducing `andMap`, I don't think we need `map4` or `map5` anymore. I think these functions can be removed to simplify the API.

File: src/Task.gren [hunk 4]
```diff
@@ -145,70 +150,38 @@ If it fails, the whole thing fails!
...
 map2 : (a -> b -> result) -> Task x a -> Task x b -> Task x result
-map2 func taskA taskB =
-    taskA
-        |> andThen
-            (\a ->
-                taskB
-                    |> andThen (\b -> succeed (func a b))
-            )
+map2 =
+    Gren.Kernel.Scheduler.map2
```
Comment (robinheghan) [inline src/Task.gren:154]: `map2` is being used in the implementation of `sequence`, which means that `sequence` is now concurrent as well.

Could you find a new implementation for `sequence` that is non-concurrent?

**Review by robinheghan — CHANGES_REQUESTED** [review]
First of all: it works! This actually speed up compiling-the-compiler a tiny amount in my very limited test :)

That said, I think there are still a few tweaks that needs to be done.

File: src/Task.gren (hunk 4 above)
Comment (andrewMacmurray) [reply src/Task.gren:154]: My bad, and well spotted! Have defined it using `andThen` now

File: src/Gren/Kernel/Scheduler.js (hunk 2 above)
Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js:173]: I agree the current implementation is half baked.

I think the problem with the one above is `_Scheduler_queue.shift` gets very slow with large arrays, might be worth using a different data structure for the queue. What do you think?

File: src/Task.gren [hunk 5]
```diff
@@ -145,70 +150,8 @@ If it fails, the whole thing fails!
...
-        |> andThen
-            (\a ->
-                taskB
-                    |> andThen (\b -> succeed (func a b))
-            )
-
-
-{-| -}
-map3 : (a -> b -> c -> result) -> Task x a -> Task x b -> Task x c -> Task x result
-map3 func taskA taskB taskC =
```
Comment (robinheghan) [inline src/Task.gren:159]: I actually thought `map3` was fine 😅

File: src/Task.gren [hunk 6]
```diff
@@ -219,8 +162,13 @@ sequence fails.
 
 -}
 sequence : Array (Task x a) -> Task x (Array a)
-sequence tasks =
-    Array.foldr (map2 Array.pushFirst) (succeed []) tasks
+sequence =
+    Array.foldr (\task combined -> task |> andThen (\x -> map (Array.pushFirst x) combined)) (succeed [])
```
Comment (robinheghan) [inline src/Task.gren:166]: Excellent 🤌

File: src/Gren/Kernel/Scheduler.js (hunk 2 above)
Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js]: I added a basic fifo queue (first commit) which speeds things up nicely, and we don't even need the separate `_Scheduler_enqueueBatch` now.

File: src/Gren/Kernel/Scheduler.js [hunk 7]
```diff
@@ -111,15 +158,15 @@ type alias Process =
 */
 
 var _Scheduler_working = false;
-var _Scheduler_queue = [];
+var _Scheduler_queue = _Utils_fifoQueue();
```
Comment (robinheghan) [inline src/Gren/Kernel/Scheduler.js:161]: Neat idea. I do think we can get away with something even simpler, though.

Instead of creating a new datastructure, we can simply reset the queue before stepping through it. That way, new tasks are added to a brand new "queue", and we don't actually need to perform a `.shift()` operation.

Like this:

```js
function _Scheduler_enqueue(proc) {
  _Scheduler_queue.push(proc);

  if (_Scheduler_working) {
    return;
  }
  _Scheduler_working = true;
  
  // Make sure tasks created during _step are run
  while (_Scheduler_queue.length > 0) {
    const activeProcs = _Scheduler_queue;
    _Scheduler_queue = [];

    for (const proc of activeProcs) {
      _Scheduler_step(proc);
    }
  }

  _Scheduler_working = false;
}
```

The downside to this compared to the original solution is that we won't clear a task for memory, until all the other tasks that were scheduled with it have run. I don't expect that to be much of a problem in practice, though.

File: src/Gren/Kernel/Scheduler.js (hunk 2 above)
Comment (robinheghan) [reply src/Gren/Kernel/Scheduler.js]: I ended up adding a new comment for the fifo queue work, so marking this discussion as resolved.

File: src/Gren/Kernel/Scheduler.js (hunk 7 above)
Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js:161]: Oh neat, that's much nicer. We can see if people run into memory problems in the future but easy to change if that's the case.

File: src/Task.gren (hunk 5 above)
Comment (andrewMacmurray) [reply src/Task.gren:159]: Oh sorry, I misread your original comment.

On a side note, do you think you'll do a similar thing with `Maybe`, `Result`, `Random` etc for consistency? They each have `map5`.

Comment (robinheghan) [reply src/Task.gren:159]: > On a side note, do you think you'll do a similar thing with ...

I've been meaning to, yes. :)

File: src/Gren/Kernel/Scheduler.js [hunk 8]
```diff
@@ -49,6 +49,47 @@ function _Scheduler_receive(callback) {
...
+        count++;
+        if (count === tasks.length) {
+          callback(_Scheduler_succeed(results));
+        }
+      }
+      const success = A2(_Scheduler_andThen, onSuccess, task);
+      const handled = A2(_Scheduler_onError, onError, success);
+      return handled;
+    });
+    procs = batch.map(_Scheduler_rawSpawn);
```
Comment (robinheghan) [inline src/Gren/Kernel/Scheduler.js:81]: nitpick: maybe perform the `rawSpawn` inside of `tasks.map` to save one iteration through the array of tasks?

Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js:81]: There was an edge case I found with this in the experiments repo I can't quite remember (something where the behaviour with an immediate failure was different) - will dig it up and write a test for it.

Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js]: So this edge case was just for the experiments repo, have added tests and refactored to this now :)

File: src/Gren/Kernel/Scheduler.js [hunk 9]
```diff
@@ -49,6 +49,47 @@ function _Scheduler_receive(callback) {
   };
 }
 
+function _Scheduler_concurrent(tasks) {
+  if (tasks.length === 0) return _Scheduler_succeed([]);
+
+  return _Scheduler_binding(function (callback) {
+    let count = 0;
+    let results = [];
```
Comment (robinheghan) [inline src/Gren/Kernel/Scheduler.js:57]: Nitpick: You can initialize this as `new Array(tasks.length)`. This should pre-initialize the array with the correct length, which I suspect will play a bit better with the JIT. Unlikely to make a huge difference in perf though.

File: src/Task.gren [hunk 10]
```diff
@@ -14,12 +14,12 @@ HTTP requests or writing to a database.
...
 ## Chains
 
-@docs andThen, await, succeed, fail, sequence
+@docs andThen, await, succeed, fail, sequence, concurrent
 
 
 ## Maps
 
-@docs map, map2, map3, map4, map5
+@docs map, andMap, map2, map3
```
Comment (robinheghan) [inline src/Task.gren:22]: Maybe `andMap` should come after `map2` and `map3`?

File: src/Task.gren [hunk 11]
```diff
@@ -125,6 +125,32 @@ map func taskA =
...
+
+    succeed ((+) 2)
+        |> andMap (succeed 3)
+    --> succeed 5
+
+    fail "oh dear"
+        |> andMap (succeed 3)
+    --> fail "oh dear"
+
+This can be used to do `Task.mapN` for any number of arguments:
```
Comment (robinheghan) [inline src/Task.gren:138]: Maybe add an explicit line that this can be helpful when `map2` or `map3` isn't enough?

File: src/Task.gren [hunk 12]
```diff
@@ -125,6 +125,32 @@ map func taskA =
...
         |> andThen (\a -> succeed (func a))
 
+{-| Apply the results of two tasks together if they succeed. Both tasks are run concurrently, and if either task fails the whole task fails:
+
+    succeed ((+) 2)
+        |> andMap (succeed 3)
+    --> succeed 5
+
+    fail "oh dear"
+        |> andMap (succeed 3)
```
Comment (robinheghan) [inline src/Task.gren:135]: Since `fail` and `succeed` can be different types, this is valid, but the type mismatch threw me off at first glance.

File: src/Task.gren [hunk 13]
```diff
@@ -219,8 +192,21 @@ sequence fails.
...
+sequence =
+    Array.foldr (\task combined -> task |> andThen (\x -> map (Array.pushFirst x) combined)) (succeed []) 
+
+
+{-| Start with an array of tasks, and turn them into a single task that returns a
+array. The tasks will be run concurrently and if any task fails the whole array fails.
+
+    concurrent [ succeed 1, succeed 2 ] == succeed [ 1, 2 ]
+
+Additionally if any task fails, any tasks already in-flight will be cleaned up (e.g. an in-flgiht http request will be cancelled).
```
Comment (robinheghan) [inline src/Task.gren:204]: I wonder if we should add a comment about why you'd ever choose `sequence` over `concurrent`? Like "If you're doing something really expensive, like making 100 http requests, you might want to do that sequentially instead of concurrently to avoid DDOSing the server" or something.

**Review by robinheghan — CHANGES_REQUESTED** [review]
I think we can do a little better on the documentation, but the implementation and the tests look good to me.

Might be worth waiting on @blaix 's input before making changes.

File: src/Gren/Kernel/Scheduler.js (hunk 9 above)
Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js:57]: I guess the same thing isn't true for `procs` as it's re-assigned in `procs = tasks.map((task, i) => { ...`?

File: src/Task.gren (hunk 12 above)
Comment (andrewMacmurray) [reply src/Task.gren:135]: Do you think it would be clearer like?:

```elm
succeed ((+) 2)
    |> andMap (succeed 3)
--> succeed 5

succeed ((+) 2)
    |> andMap (fail "oh dear")
--> fail "oh dear"
```

File: src/Task.gren (hunk 13 above)
Comment (andrewMacmurray) [reply src/Task.gren:204]: I like that.

Do you think it's worth adding a note here about `concurrent` vs `parallel`? Because JS is single threaded, `concurrent` won't speed up CPU bound tasks, but IO bound tasks like http may see a speed boost.

Would need some friendlier wording and examples but something along those lines.

Comment (robinheghan) [reply src/Task.gren:204]: Yeah, I think that would be good.

**Comment by robinheghan** [comment]
Would you mind adding your name to `CONTRIBUTORS` as well?

File: src/Gren/Kernel/Scheduler.js (hunk 9 above)
Comment (andrewMacmurray) [reply src/Gren/Kernel/Scheduler.js:57]: We could do:
```
let procs
```
Which initialises to `undefined`

File: src/Task.gren [hunk 14]
```diff
@@ -219,8 +194,34 @@ sequence fails.
...
+
+**Note**: Why use `sequence` over `concurrent`? Maybe you have an expensive operation, like making 100 http requests.
+You might want to do these sequentially rather than all at once to avoid DDOSing the server.
```
Comment (robinheghan) [inline src/Task.gren:209]: Maybe DDOSing is a too technical term? «to avoid overloading the server» might be better? 🤔

**Comment by robinheghan** [comment]
Thank you for your hard work on this, @andrewMacmurray <3
